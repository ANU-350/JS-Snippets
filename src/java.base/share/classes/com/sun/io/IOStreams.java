/*
 * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package com.sun.io;

import java.io.ByteArrayOutputStream;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.io.PipedOutputStream;
import java.util.zip.CheckedOutputStream;

/**
 * Package-private utility class
 */
final class IOStreams {

    /**
     * Returns true if this class satisfies two conditions:
     * <pre>
     * - the reference to {@code byte[]} is not kept within the class
     * - the argument of {@link OutputStream#write(byte[])}} and {@link OutputStream#write(byte[], int, int)}} is not modified within the methods
     * - the {@code byte[]} is not read outside of the given bounds
     * </pre>
     *
     * @return true if this class is trusted
     * @see java.io.ByteArrayInputStream#transferTo(OutputStream)
     * @see java.io.BufferedInputStream#implTransferTo(OutputStream)
     */
    public static boolean isTrusted(OutputStream os) {
        var clazz = os.getClass();
        return clazz == ByteArrayOutputStream.class
                || clazz == FileOutputStream.class
                || clazz == PipedOutputStream.class
                || clazz == CheckedOutputStream.class;
    }
}
