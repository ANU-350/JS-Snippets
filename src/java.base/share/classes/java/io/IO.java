/*
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.io;

import jdk.internal.javac.PreviewFeature;

/**
 * A collection of static convenience methods that provide access to
 * {@link System#console()} for implicitly declared classes.
 *
 * <p> Each of this class' methods calls a similarly-named method on
 * {@code Console} returned by {@code System.console()}, or throws
 * {@link IOError} if {@code System.console()} returns {@code null}.
 *
 * @since 23
 */
@PreviewFeature(feature = PreviewFeature.Feature.IMPLICIT_CLASSES)
public class IO {

    private IO() {
        throw new Error("no instances");
    }

    /**
     * Calls {@link Console#println(Object) Console.println(obj)} on
     * {@link System#console()}, or throws {@link IOError} if
     * {@code System.console()} returns {@code null}.
     *
     * @param obj the object to print
     */
    public static void println(Object obj) {
        con().println(obj);
    }

    /**
     * Calls {@link Console#print(Object) Console.print(obj)} on
     * {@link System#console()}, or throws {@link IOError} if
     * {@code System.console()} returns {@code null}.
     *
     * @param obj the object to print
     */
    public static void print(Object obj) {
        con().print(obj);
    }

    /**
     * {@return the result of a call to {@link Console#readln(String)
     * Console.readln(prompt)} on {@link System#console()}, or throws
     * {@link IOError} if {@code System.console()} returns {@code null}}
     *
     * @param prompt the prompt string
     */
    public static String readln(String prompt) {
        return con().readln(prompt);
    }

    private static Console con() {
        var con = System.console();
        if (con != null) {
            return con;
        } else {
            throw new IOError(null);
        }
    }
}
