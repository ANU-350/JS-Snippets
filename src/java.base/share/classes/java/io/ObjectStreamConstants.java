/*
 * Copyright (c) 1996, 2023, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.io;

/**
 * Constants written into the Object Serialization Stream.
 *
 * @since 1.1
 */
public interface ObjectStreamConstants {

    /**
     * Magic number that is written to the stream header.
     */
    static short STREAM_MAGIC = (short)0xaced;

    /**
     * Version number that is written to the stream header.
     */
    static short STREAM_VERSION = 5;

    /* Each item in the stream is preceded by a tag
     */

    /**
     * First tag value.
     */
    static byte TC_BASE = 0x70;

    /**
     * Null object reference.
     */
    static byte TC_NULL = 0x70;

    /**
     * Reference to an object already written into the stream.
     */
    static byte TC_REFERENCE = 0x71;

    /**
     * new Class Descriptor.
     */
    static byte TC_CLASSDESC = 0x72;

    /**
     * new Object.
     */
    static byte TC_OBJECT = 0x73;

    /**
     * new String.
     */
    static byte TC_STRING = 0x74;

    /**
     * new Array.
     */
    static byte TC_ARRAY = 0x75;

    /**
     * Reference to Class.
     */
    static byte TC_CLASS = 0x76;

    /**
     * Block of optional data. Byte following tag indicates number
     * of bytes in this block data.
     */
    static byte TC_BLOCKDATA = 0x77;

    /**
     * End of optional block data blocks for an object.
     */
    static byte TC_ENDBLOCKDATA = 0x78;

    /**
     * Reset stream context. All handles written into stream are reset.
     */
    static byte TC_RESET = 0x79;

    /**
     * long Block data. The long following the tag indicates the
     * number of bytes in this block data.
     */
    static byte TC_BLOCKDATALONG = 0x7A;

    /**
     * Exception during write.
     */
    static byte TC_EXCEPTION = 0x7B;

    /**
     * Long string.
     */
    static byte TC_LONGSTRING = 0x7C;

    /**
     * new Proxy Class Descriptor.
     */
    static byte TC_PROXYCLASSDESC = 0x7D;

    /**
     * new Enum constant.
     * @since 1.5
     */
    static byte TC_ENUM = 0x7E;

    /**
     * Last tag value.
     */
    static byte TC_MAX = 0x7E;

    /**
     * First wire handle to be assigned.
     */
    static int baseWireHandle = 0x7e0000;


    /* *******************************************************************/
    /* Bit masks for ObjectStreamClass flag. */

    /**
     * Bit mask for ObjectStreamClass flag. Indicates a Serializable class
     * defines its own writeObject method.
     */
    static byte SC_WRITE_METHOD = 0x01;

    /**
     * Bit mask for ObjectStreamClass flag. Indicates Externalizable data
     * written in Block Data mode.
     * Added for PROTOCOL_VERSION_2.
     *
     * @see #PROTOCOL_VERSION_2
     * @since 1.2
     */
    static byte SC_BLOCK_DATA = 0x08;

    /**
     * Bit mask for ObjectStreamClass flag. Indicates class is Serializable.
     */
    static byte SC_SERIALIZABLE = 0x02;

    /**
     * Bit mask for ObjectStreamClass flag. Indicates class is Externalizable.
     */
    static byte SC_EXTERNALIZABLE = 0x04;

    /**
     * Bit mask for ObjectStreamClass flag. Indicates class is an enum type.
     * @since 1.5
     */
    static byte SC_ENUM = 0x10;


    /* *******************************************************************/
    /* Security permissions */

    /**
     * Enable substitution of one object for another during
     * serialization/deserialization.
     *
     * @see java.io.ObjectOutputStream#enableReplaceObject(boolean)
     * @see java.io.ObjectInputStream#enableResolveObject(boolean)
     * @since 1.2
     */
    static SerializablePermission SUBSTITUTION_PERMISSION =
                           new SerializablePermission("enableSubstitution");

    /**
     * Enable overriding of readObject and writeObject.
     *
     * @see java.io.ObjectOutputStream#writeObjectOverride(Object)
     * @see java.io.ObjectInputStream#readObjectOverride()
     * @since 1.2
     */
    static SerializablePermission SUBCLASS_IMPLEMENTATION_PERMISSION =
                    new SerializablePermission("enableSubclassImplementation");

    /**
     * Enable setting the system-wide serial filter.
     *
     * @see java.io.ObjectInputFilter.Config#setSerialFilter(ObjectInputFilter)
     * @since 9
     */
    static SerializablePermission SERIAL_FILTER_PERMISSION =
            new SerializablePermission("serialFilter");

   /**
    * A Stream Protocol Version. <p>
    *
    * All externalizable data is written in JDK 1.1 external data
    * format after calling this method. This version is needed to write
    * streams containing Externalizable data that can be read by
    * pre-JDK 1.1.6 JVMs.
    *
    * @see java.io.ObjectOutputStream#useProtocolVersion(int)
    * @since 1.2
    */
    static int PROTOCOL_VERSION_1 = 1;


   /**
    * A Stream Protocol Version. <p>
    *
    * This protocol is written by JVM 1.2.
    * <p>
    * Externalizable data is written in block data mode and is
    * terminated with TC_ENDBLOCKDATA. Externalizable class descriptor
    * flags has SC_BLOCK_DATA enabled. JVM 1.1.6 and greater can
    * read this format change.
    * <p>
    * Enables writing a nonSerializable class descriptor into the
    * stream. The serialVersionUID of a nonSerializable class is
    * set to 0L.
    *
    * @see java.io.ObjectOutputStream#useProtocolVersion(int)
    * @see #SC_BLOCK_DATA
    * @since 1.2
    */
    static int PROTOCOL_VERSION_2 = 2;
}
