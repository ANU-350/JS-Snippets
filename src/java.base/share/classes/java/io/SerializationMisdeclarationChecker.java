/*
 * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.io;

import java.lang.reflect.Field;
import java.lang.reflect.Member;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;
import java.security.AccessController;
import java.security.PrivilegedAction;
import java.util.Arrays;

import static jdk.internal.event.SerializationMisdeclarationEvent.*;
import static java.lang.reflect.Modifier.*;

final class SerializationMisdeclarationChecker {

    private static final String SUID_NAME = "serialVersionUID";
    private static final String SERIAL_PERSISTENT_FIELDS_NAME = "serialPersistentFields";
    private static final String WRITE_OBJECT_NAME = "writeObject";
    private static final String READ_OBJECT_NAME = "readObject";
    private static final String READ_OBJECT_NO_DATA_NAME = "readObjectNoData";
    private static final String WRITE_REPLACE_NAME = "writeReplace";
    private static final String READ_RESOLVE_NAME = "readResolve";

    private static final Class<?>[] WRITE_OBJECT_PARAM_TYPES = {ObjectOutputStream.class};
    private static final Class<?>[] READ_OBJECT_PARAM_TYPES = {ObjectInputStream.class};
    private static final Class<?>[] READ_OBJECT_NO_DATA_PARAM_TYPES = {};
    private static final Class<?>[] WRITE_REPLACE_PARAM_TYPES = {};
    private static final Class<?>[] READ_RESOLVE_PARAM_TYPES = {};

    private final Class<?> cl;

    SerializationMisdeclarationChecker(Class<?> cl) {
        this.cl = cl;
    }

    @SuppressWarnings("removal")
    void checkMisdeclarations() {
        AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
            privilegedCheckSerialVersionUID();
            privilegedCheckSerialPersistentFields();

            privilegedCheckPrivateMethod(WRITE_OBJECT_NAME, WRITE_OBJECT_PARAM_TYPES, Void.TYPE);
            privilegedCheckPrivateMethod(READ_OBJECT_NAME, READ_OBJECT_PARAM_TYPES, Void.TYPE);
            privilegedCheckPrivateMethod(READ_OBJECT_NO_DATA_NAME, READ_OBJECT_NO_DATA_PARAM_TYPES, Void.TYPE);

            privilegedCheckAccessibleMethod(WRITE_REPLACE_NAME, WRITE_REPLACE_PARAM_TYPES, Object.class);
            privilegedCheckAccessibleMethod(READ_RESOLVE_NAME, READ_RESOLVE_PARAM_TYPES, Object.class);

            return null;
        });
    }

    private void privilegedCheckSerialVersionUID() {
        Field f = declaredField(cl, SUID_NAME);
        if (f == null) {
            if (isOrdinaryClass()) {
                commitEvent(SUID_EXPLICIT,
                        "a private static final long " + SUID_NAME +
                                " field should be declared");
            }
            return;
        }
        if (cl.isEnum()) {
            commitEvent(SUID_INEFFECTIVE_ENUM,
                    SUID_NAME + " is not effective in an enum class");
        }
        if (!isPrivate(f)) {
            commitEvent(SUID_PRIVATE,
                    SUID_NAME + " should be declared private");
        }
        if (!isStatic(f)) {
            commitEvent(SUID_STATIC,
                    SUID_NAME + " must be declared static to be effective");
        }
        if (!isFinal(f)) {
            commitEvent(SUID_FINAL,
                    SUID_NAME + " must be declared final to be effective");
        }
        if (f.getType() != Long.TYPE) {
            commitEvent(SUID_LONG,
                    SUID_NAME + " should be declared of type long");
        }
        if (!isStatic(f)) {
            return;
        }
        f.setAccessible(true);
        if (getLong(f) == null) {
            commitEvent(SUID_CONVERTIBLE_TO_LONG,
                    SUID_NAME + " must be convertible to long via widening to be effective");
        }
    }

    private void privilegedCheckSerialPersistentFields() {
        Field f = declaredField(cl, SERIAL_PERSISTENT_FIELDS_NAME);
        if (f == null) {
            return;
        }
        if (cl.isRecord()) {
            commitEvent(SER_PERS_INEFFECTIVE_RECORD,
                    SERIAL_PERSISTENT_FIELDS_NAME + " is not effective in a record class");
        } else if (cl.isEnum()) {
            commitEvent(SER_PERS_INEFFECTIVE_ENUM,
                    SERIAL_PERSISTENT_FIELDS_NAME + " is not effective in an enum class");
        }
        if (!isPrivate(f)) {
            commitEvent(SER_PERS_PRIVATE,
                    SERIAL_PERSISTENT_FIELDS_NAME + " must be declared private to be effective");
        }
        if (!isStatic(f)) {
            commitEvent(SER_PERS_STATIC,
                    SERIAL_PERSISTENT_FIELDS_NAME + " must be declared static to be effective");
        }
        if (!isFinal(f)) {
            commitEvent(SER_PERS_FINAL,
                    SERIAL_PERSISTENT_FIELDS_NAME + " must be declared final to be effective");
        }
        if (f.getType() != ObjectStreamField[].class) {
            commitEvent(SER_PERS_TYPE_OSF_ARRAY,
                    SERIAL_PERSISTENT_FIELDS_NAME + " should be declared of type ObjectStreamField[]");
        }
        if (!isStatic(f)) {
            return;
        }
        f.setAccessible(true);
        Object spf = getObject(f);
        if (spf == null) {
            commitEvent(SER_PERS_NOT_NULL,
                    SERIAL_PERSISTENT_FIELDS_NAME + " must not be null to be effective");
            return;
        }
        if (!(spf instanceof ObjectStreamField[])) {
            commitEvent(SER_PERS_VALUE_OSF_ARRAY,
                    SERIAL_PERSISTENT_FIELDS_NAME + " must be an instance of ObjectStreamField[] to be effective");
        }
    }

    private void privilegedCheckPrivateMethod(String name, Class<?>[] paramTypes, Class<?> retType) {
        for (Method m : cl.getDeclaredMethods()) {
            if (m.getName().equals(name)) {
                privilegedCheckPrivateMethod(m, paramTypes, retType);
            }
        }
    }

    private void privilegedCheckPrivateMethod(Method m, Class<?>[] paramTypes, Class<?> retType) {
        if (cl.isEnum()) {
            commitEvent(PRIV_METH_INEFFECTIVE_ENUM,
                    m + " is not effective on an enum class");
        } else if (cl.isRecord()) {
            commitEvent(PRIV_METH_INEFFECTIVE_RECORD,
                    m + " is not effective on a record class");
        }
        if (!isPrivate(m)) {
            commitEvent(PRIV_METH_PRIV,
                    m + " must be private to be effective");
        }
        if (isStatic(m)) {
            commitEvent(PRIV_METH_NON_STATIC,
                    m + " must be non-static to be effective");
        }
        if (m.getReturnType() != retType) {
            commitEvent(PRIV_METH_RET_TYPE,
                    m + " must have return type " + retType + " to be effective");
        }
        if (!Arrays.equals(m.getParameterTypes(), paramTypes)) {
            commitEvent(PRIV_METH_PARAM_TYPES,
                    m + " must have parameter types " + Arrays.toString(paramTypes) + " to be effective");
        }
    }

    private void privilegedCheckAccessibleMethod(String name, Class<?>[] paramTypes, Class<?> retType) {
        for (Class<?> cls = cl; cls != null; cls = cls.getSuperclass()) {
            for (Method m : cls.getDeclaredMethods()) {
                if (m.getName().equals(name)) {
                    privilegedCheckAccessibleMethod(cls, m, paramTypes, retType);
                }
            }
        }
    }

    private void privilegedCheckAccessibleMethod(Class<?> cls, Method m, Class<?>[] paramTypes, Class<?> retType) {
        if (cls.isEnum()) {
            commitEvent(cls, ACC_METH_INEFFECTIVE_ENUM,
                    m + " is not effective on an enum class");
        }
        if (isAbstract(m)) {
            commitEvent(ACC_METH_NON_ABSTRACT,
                    m + " must be non-abstract to be effective");
        }
        if (isStatic(m)) {
            commitEvent(ACC_METH_NON_STATIC,
                    m + " must be non-static to be effective");
        }
        if (m.getReturnType() != retType) {
            commitEvent(ACC_METH_RET_TYPE,
                    m + " must have return type " + retType + " to be effective");
        }
        if (!Arrays.equals(m.getParameterTypes(), paramTypes)) {
            commitEvent(ACC_METH_PARAM_TYPES,
                    m + " must have parameter types " + Arrays.toString(paramTypes) + " to be effective");
        }
        if (isPrivate(m) && cl != cls
                || isPackageProtected(m) && !isSamePackage(cl, cls)) {
            commitEvent(cl, ACC_METH_NON_ACCESSIBLE,
                    m + " is not accessible from " + cl);
        }
    }

    private static boolean isSamePackage(Class<?> cl, Class<?> mcl) {
        return cl.getClassLoader() == mcl.getClassLoader()
                && cl.getPackageName().equals(mcl.getPackageName());
    }

    private boolean isOrdinaryClass() {
        /* class Enum and class Record are not considered ordinary classes */
        return !(cl.isRecord() || cl.isEnum() || cl.isArray()
                || Enum.class == cl || Record.class == cl
                || Proxy.isProxyClass(cl));
    }

    private static boolean isPrivate(Member m) {
        return (m.getModifiers() & PRIVATE) != 0;
    }

    private static boolean isPackageProtected(Member m) {
        return (m.getModifiers()
                & (PRIVATE | PROTECTED | PUBLIC)) == 0;
    }

    private static boolean isAbstract(Member m) {
        return (m.getModifiers() & ABSTRACT) != 0;
    }

    private static boolean isFinal(Member m) {
        return (m.getModifiers() & FINAL) != 0;
    }

    private static boolean isStatic(Member m) {
        return (m.getModifiers() & STATIC) != 0;
    }

    private static Field declaredField(Class<?> cl, String name) {
        try {
            return cl.getDeclaredField(name);
        } catch (NoSuchFieldException e) {
            return null;
        }
    }

    private static Object getObject(Field f) {
        try {
            return f.get(null);
        } catch (IllegalAccessException ignored) {
        }
        return null;
    }

    private static Long getLong(Field f) {
        try {
            return f.getLong(null);
        } catch (IllegalArgumentException | IllegalAccessException ignored) {
        }
        return null;
    }

    private void commitEvent(int kind, String msg, Object... args) {
        commitEvent(cl, kind, msg);
    }

    private static void commitEvent(Class<?> cls, int kind, String msg) {
        commit(timestamp(), 0, cls, kind, msg);
    }

}
