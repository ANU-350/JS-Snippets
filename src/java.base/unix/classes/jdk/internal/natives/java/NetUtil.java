package jdk.internal.natives.java;

public final class NetUtil {

    private NetUtil() {
    }

    // Todo: Implement this method properly. See net_util.c
    public static boolean ipv6_available() {
        return true;
    }

}
