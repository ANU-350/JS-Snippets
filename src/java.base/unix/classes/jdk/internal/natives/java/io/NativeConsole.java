/*
 * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package jdk.internal.natives.java.io;

import java.lang.invoke.MethodHandle;

import static java.lang.foreign.ValueLayout.JAVA_INT;
import static jdk.internal.foreign.support.LookupUtil.*;
import static jdk.internal.foreign.support.InvokeUtil.newInternalError;

public final class NativeConsole {

    private NativeConsole() {}

    public static boolean istty() {
        return isatty(0) && isatty(1);
    }

    public static String encoding() {
        return null;
    }

    // Native methods

    // https://man7.org/linux/man-pages/man3/isatty.3.html
    private static final MethodHandle IS_A_TTY = downcall("isatty", JAVA_INT, JAVA_INT);

    static boolean isatty(int fd) {
        try {
            return (int) IS_A_TTY.invokeExact(fd) == 1;
        } catch (Throwable t) {
            throw newInternalError(IS_A_TTY, t);
        }
    }
}
