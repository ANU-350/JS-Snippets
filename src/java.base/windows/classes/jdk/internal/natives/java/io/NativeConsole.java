/*
 * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package jdk.internal.natives.java.io;

import java.lang.foreign.MemorySegment;
import java.lang.invoke.MethodHandle;
import java.nio.charset.StandardCharsets;

import static jdk.internal.foreign.support.DefaultNativeLookupUtil.downcall;
import static jdk.internal.foreign.support.DefaultNativeLookupUtil.downcallOfVoid;
import static jdk.internal.foreign.support.InvokeUtil.invokeAsInt;
import static jdk.internal.natives.WindowsConstants.*;
import static jdk.internal.natives.WindowsMethods.getFileType;
import static jdk.internal.natives.WindowsMethods.getStdHandle;
import static jdk.internal.natives.WindowsTypes.*;

public final class NativeConsole {

    private NativeConsole() {}

    public static boolean istty() {
        MemorySegment hStdIn = getStdHandle(STD_INPUT_HANDLE);
        MemorySegment hStdOut = getStdHandle(STD_OUTPUT_HANDLE);

        return !isInvalidHandleValue(hStdIn) &&
                !isInvalidHandleValue(hStdOut) &&
                getFileType(hStdIn) == FILE_TYPE_CHAR &&
                getFileType(hStdOut) == FILE_TYPE_CHAR;

    }

    public static String encoding() {
        int cp = invokeAsInt(GET_CONSOLE_CP);
        if (cp >= 874 && cp <= 950) {
            return "ms" + cp;
        }
        if (cp == 65001) {
            return StandardCharsets.UTF_8.name();
        }
        return "cp" + cp;
    }

    // Native methods

    // https://learn.microsoft.com/en-us/windows/console/getconsolecp
    private static final MethodHandle GET_CONSOLE_CP = downcall("GetConsoleCP", UINT);

}
