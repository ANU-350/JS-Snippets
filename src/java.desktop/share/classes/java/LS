-rw-r--r--   1 prrace  staff   22640 Sep 21 12:12 CheckboxMenuItem.java
-rw-r--r--   1 prrace  staff    8599 Sep 21 12:12 PopupMenu.java
-rw-r--r--   1 prrace  staff    4344 Sep 21 12:12 Panel.java
-rw-r--r--   1 prrace  staff    9093 Sep 21 12:11 Canvas.java
-rw-r--r--   1 prrace  staff   25224 Sep 21 12:10 TextArea.java
-rw-r--r--   1 prrace  staff   30118 Sep 21 12:10 ScrollPane.java
-rw-r--r--   1 prrace  staff   48362 Sep 21 12:09 Scrollbar.java
-rw-r--r--   1 prrace  staff   27764 Sep 21 12:08 TextField.java
-rw-r--r--   1 prrace  staff   23182 Sep 21 12:08 Button.java
-rw-r--r--   1 prrace  staff   20371 Sep 21 12:07 Menu.java
-rw-r--r--   1 prrace  staff   31976 Sep 21 12:07 MenuItem.java
-rw-r--r--   1 prrace  staff   16430 Sep 21 12:06 MenuBar.java
-rw-r--r--   1 prrace  staff   45588 Sep 21 12:06 Frame.java
-rw-r--r--   1 prrace  staff   64543 Sep 21 12:01 Dialog.java
-rw-r--r--   1 prrace  staff  154234 Sep 21 12:00 Window.java
-rw-r--r--   1 prrace  staff  195174 Sep 21 11:59 Container.java
-rw-r--r--   1 prrace  staff  390777 Sep 21 11:57 Component.java
