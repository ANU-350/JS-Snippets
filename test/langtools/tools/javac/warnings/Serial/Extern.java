/*
 * @test /nodynamiccopyright/
 * @bug 8202056
 * @compile/ref=Extern.out -XDrawDiagnostics -Xlint:serial Extern.java
 */


import java.io.*;

class Extern implements Externalizable {
    private static final long serialVersionUID = 42;

    // ineffectual
    private static final ObjectStreamField[] serialPersistentFields = {};

    public void writeExternal(ObjectOutput out) throws IOException {
	return;
    }

    public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
	return;
    }

    // ineffectual
    private void readObject(ObjectInputStream stream)
	throws IOException, ClassNotFoundException {
	return;
    }

    // ineffectual
    private void writeObject(ObjectOutputStream stream) throws IOException {
	return;
    }

    // ineffectual
    private void readObjectNoData() throws ObjectStreamException {
	return;
    }

    // (possibly) effective
    private Object writeReplace() throws ObjectStreamException {
	return null;
    }

    // (possibly) effective
    private Object readResolve() throws ObjectStreamException {
	return null;
    }

}
