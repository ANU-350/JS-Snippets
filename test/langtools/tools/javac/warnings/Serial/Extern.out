Extern.java:10:1: compiler.warn.ineffectual.serial.field.externalizable
Extern.java:10:1: compiler.warn.ineffectual.serial.method.externalizable: readObject(java.io.ObjectInputStream)
Extern.java:10:1: compiler.warn.ineffectual.serial.method.externalizable: writeObject(java.io.ObjectOutputStream)
Extern.java:10:1: compiler.warn.ineffectual.serial.method.externalizable: readObjectNoData()
4 warnings
